### DISCUSSION TITLE
Apollo Dev tools as Flipper Plugin

### DISCUSSION THREAD

**Discussion by hwillson** [apollographql/apollo-client-devtools#385] [Ideas]
Migrated from: https://github.com/apollographql/apollo-client-devtools/issues/255

React Native dev tooling is going to Flipper more and more. Flipper seems to have a good plugin system. It would be really nice to have the Apollo dev tools as a flipper plugin. I have no idea how feasible this would be, but would be interested in taking a look at it if this is thought a good idea & someone can help point out to right direction.

**Comment by hsavit1** [comment]
I've used the Apollo-dev-tools flipper plugin v0.2 - works decently well as a proof of concept, but viewing queries and mutations do not work, only cache inspection. 

Anyway, I'm 100% confident that this would make any mobile developer's life more productive

**Comment by tvler** [comment]
I think this is a great idea and would benefit the ecosystem a ton

**Comment by gentlee** [comment]
So what's the status? We need it a lot. If you have reanimated lib in your react-native project you are tied to Flipper and debugging apollo is a real pain.

**Comment by Yasir5247** [comment]
I think this is a great idea too

**Comment by aq1018** [comment]
We need an **Official** version of flipper client. The unofficial version is unmaintained and has many features missing. 

We have been a long time Apollo user. We love Apollo Studio and how easy things just work with Apollo. We value Apollo precisely because of the developer friendliness. I happen to believe this is what Apollo believes in as well, developer friendliness. 

But things are not as easy and painless here on React Native side. It would be awesome to improve the situation here. More and more companies are moving to React Native and using Flipper as THE Developer Tool for debugging. This is because Flipper is included by default after RN 0.62.

**Comment by trymbill** [comment]
This is definitely something Apollo should look into and give some serious thought. Having an official plugin in Flipper for Apollo would be a massive sticky point for us to continue working with Apollo in our projects.

**Comment by sammarks** [comment]
I wonder since Flipper exposes the Hermes Debugger, which is basically just an instance of Chrome DevTools, maybe it would be an "easier" route to somehow install the Apollo DevTools extension into that first rather than having to re-package as a Flipper plugin.

**Comment by mohit23x** [comment]
Hey folks, presenting you the apollo client devtools as a flipper plugin to support React Native Apps. Check it out [here](https://github.com/razorpay/react-native-apollo-devtools)

**Reply by declanelcocks** [reply]
Looks great @mohit23x 👍

**Reply by shubhnik** [reply]
Nice work @mohit23x 💯

**Reply by mohit23x** [reply]
sharing the blog which describes our journey of creating this tool, read it [here](https://engineering.razorpay.com/apollo-client-devtools-for-flipper-bc5337b3a411)

**Reply by declanelcocks** [reply]
@mohit23x after using it for a couple of weeks, my only suggestion would be include some transparency on how queries/mutations are also cached with references.

https://www.apollographql.com/docs/react/caching/advanced-topics/#cache-redirects I was trying to explain to a colleague why we had to add a cache redirect in our app for something not too dissimilar to the Apollo docs (`getBooks` + `getBook(id)`). I forget exactly how it's stored in Apollo, but I remember that `getBook` will be stored with a reference to a specific item in the cache.

Don't get me wrong, the plugin is already great as an intro to Apollo's caching and to have some transparency on what's going on when you make a query, I'm just nitpicking for future enhancements 😄

**Reply by mohit23x** [reply]
@declanelcocks The cached data shown is query is schema which was cached when the response of that query was resolved, the behaviour is same as that of apollo devtool web plugin.
